lib/http2/debug_state.c

```diff
@@ -0,0 +1,217 @@
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    } else {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+            return debug_state_string_open;
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
+{
+    char check[1]; /* for only checking the size of output */
+    va_list args;
+
+    va_start(args, fmt);
+    int size = vsnprintf(check, 1, fmt, args);
+    va_end(args);
+
+    h2o_iovec_t v;
+    v.base = h2o_mem_alloc_pool(pool, size + 1);
+
+    va_start(args, fmt);
+    v.len = vsnprintf(v.base, size + 1, fmt, args);
+    va_end(args);
+
+    lines->entries[lines->size++] = v;
+}
+
+static void append_header_table_lines(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, h2o_hpack_header_table_t *header_table)
+{
+    int i, comma_removed = 0;
+    for (i = 0; i < header_table->num_entries; i++) {
+        h2o_hpack_header_table_entry_t *entry = h2o_hpack_header_table_get(header_table, i);
+        append_line(pool, lines,
+                    ",\n"
+                    "      [ \"%.*s\", \"%.*s\" ]",
+                    (int)entry->name->len, entry->name->base,
+                    (int)entry->value->len, entry->value->base);
+        if (lines->entries[lines->size - 1].len > 0 && !comma_removed) {
+            lines->entries[lines->size - 1].base[0] = ' ';
+            comma_removed = 1;
+        }
+    }
+}
+
+h2o_debug_state_t *h2o_http2_get_debug_state(h2o_req_t *req, int hpack_enabled)
+{
+    h2o_http2_conn_t *conn = (h2o_http2_conn_t*)req->conn;
+    struct st_h2o_debug_state_t *state = h2o_mem_alloc_pool(&req->pool, sizeof(*state));
+    *state = (struct st_h2o_debug_state_t){{NULL}};
+
+    state->conn_flow_in = conn->_write.window._avail;
+    state->conn_flow_out = conn->_write.window._avail;
+
+    /* calculate total number of JSON lines */
+    size_t nr_resp = 3;
+    nr_resp += conn->num_streams.pull.open + conn->num_streams.push.open; // streams
+    if (hpack_enabled) {
+        nr_resp += 3 + conn->_input_header_table.num_entries + conn->_output_header_table.num_entries; // hpack
+    }
+    h2o_vector_reserve(&req->pool, &state->json, nr_resp);
```

Comment:
Let's call h2o_vector_reserve in `append_line` instead of trying to determine the correct length of the output here.

This is not a performance sensitive handler, and minimizing the risks of having issues should be given higher priority than optimization.